RileyLinkKit/PumpOpsSynchronous.swift

```diff
@@ -483,38 +421,71 @@ class PumpOpsSynchronous {
+        // Exceptions are Square Wave boluses, which can be out of order in the pump history by up
+        // to 8 hours on older pumps, and Normal Boluses, which can be out of order by roughly 4 minutes.
+        let eventTimestampDeltaAllowance: TimeInterval
+        if pumpModel.appendsSquareWaveToHistoryOnStartOfDelivery {
+            eventTimestampDeltaAllowance = TimeInterval(minutes: 10)
+        } else {
+            eventTimestampDeltaAllowance = TimeInterval(hours: 9)
+        }
+        
+        
+        // Start with some time in the future, to account for the condition when the pump's clock is ahead
+        // of ours by a small amount.
+        var timeCursor = Date(timeIntervalSinceNow: TimeInterval(minutes: 60))
+        var events = [TimestampedHistoryEvent]()
+        var timeAdjustmentInterval: TimeInterval = 0
+        var seenEventData = Set<Data>()
+        var lastEvent: PumpEvent?
+
+
+        for event in pumpEvents {
+            if let event = event as? TimestampedPumpEvent, !seenEventData.contains(event.rawData) {
+                seenEventData.insert(event.rawData)
+
+                var timestamp = event.timestamp
+                timestamp.timeZone = pump.timeZone
+
+                if let date = timestamp.date?.addingTimeInterval(timeAdjustmentInterval) {
+                    if date.timeIntervalSince(startDate) < -eventTimestampDeltaAllowance {
+                        NSLog("Found event at (%@) to be more than %@s before startDate(%@)", date as NSDate, String(describing: eventTimestampDeltaAllowance), startDate as NSDate);
+                        return (events, true)
+                    } else if date.timeIntervalSince(timeCursor) > eventTimestampDeltaAllowance {
+                        NSLog("Found event (%@) out of order in history. Ending history fetch.", date as NSDate)
+                        return (events, true)
+                    } else {
+                        if (date.compare(startDate) != .orderedAscending) {
+                            timeCursor = date
                         }
+                        events.insert(TimestampedHistoryEvent(pumpEvent: event, date: date), at: 0)
                     }
                 }
+            }
 
-                if let changeTimeEvent = event as? ChangeTimePumpEvent, let newTimeEvent = lastEvent as? NewTimePumpEvent {
-                    timeAdjustmentInterval += (newTimeEvent.timestamp.date?.timeIntervalSince(changeTimeEvent.timestamp.date!))!
-                }
-
-                lastEvent = event
+            if let changeTimeEvent = event as? ChangeTimePumpEvent, let newTimeEvent = lastEvent as? NewTimePumpEvent {
+                timeAdjustmentInterval += (newTimeEvent.timestamp.date?.timeIntervalSince(changeTimeEvent.timestamp.date!))!
             }
+
+            lastEvent = event
         }
-        return (events, pumpModel)
+        
+        return (events, false)
```

Comment:
The return tuple of (events: [TimestampedHistoryEvent], cancelled: Bool) would be clearer if referred to with its label names.  And please change the 'canceled' label to a positive form, like 'shouldContinue', or 'moreEvents'.